apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsLinkControl.tsx

```diff
@@ -89,24 +205,78 @@ const RadioCard = ({ children, ...rest }: PropsWithChildren<UseRadioProps>) => {
 
 interface RadioContentProps {
   selectedLinkType: string
+  description?: string
   data: string
   handleChange: (value: string) => void
 }
 
 const RadioContent = ({
   selectedLinkType,
+  description,
   data,
   handleChange,
 }: RadioContentProps): JSX.Element => {
+  const router = useRouter()
+  const siteId = String(router.query.siteId)
+  const {
+    isOpen: isPageLinkModalOpen,
+    onOpen: onPageLinkModalOpen,
+    onClose: onPageLinkModalClose,
+  } = useDisclosure()
+  const potentialInternalLinkResourceId = getResourceIdFromReferenceLink(data)
+  const { data: potentialInternalLinkResource } =
+    trpc.resource.getWithFullPermalink.useQuery({
+      resourceId: potentialInternalLinkResourceId,
+    })
+
   switch (selectedLinkType) {
     case "page":
       return (
-        <Input
-          type="text"
-          value={data}
-          onChange={(e) => handleChange(e.target.value)}
-          placeholder="Page permalink"
-        />
+        <>
+          <PageLinkModal
+            data={data}
+            description={description}
+            isOpen={isPageLinkModalOpen}
+            onClose={onPageLinkModalClose}
+            onSave={(value) => {
+              const referenceId = `[resource:${siteId}:${value}]`
```

Comment:
use `getReferenceLink` util?

Reply:
Oops, updated in 1a92e6aab9d260070adc42fafda564b266228d9f.